src/move_to_front.rs

```diff
@@ -36,8 +36,137 @@ impl MoveToFrontDecoder {
+        }
+    }
+
+    fn reset(&mut self) {
+        self.selectors.clear();
+        self.mtf_freq.clear();
+    }
+
+    // Tighly coupled with the overall compression routine
+    // as it requires the ptr array
+    // However, I find it difficult to decouple this without sacrificing performance
+    pub fn encode(
+        &mut self,
+        nblock: usize,
+        buf: &[u8],
+        ptr: &[usize],
+        in_use: &[bool; Self::BYTE_RANGE],
+    ) {
+        self.reset();
+        // TODO: reserve space for at least nblock elements
+        // not "additional" elements as the docs say
+        // self.selectors.reserve_exact();
+
+        let mut count_used_bytes = 0;
+        let mut index_selector = [0; Self::BYTE_RANGE];
+        for byte in 0..Self::BYTE_RANGE {
+            if in_use[byte] {
+                index_selector[byte] = count_used_bytes;
+                count_used_bytes += 1;
+            }
+        }
+
+        self.mtf_freq = vec![0; count_used_bytes + 2]; // +1 for EOB, +1 for 1-offset
+
+        let mut live_mtf: Vec<usize> = (0..count_used_bytes).collect();
+        let mut index_zero_freq = 0;
+
+        macro_rules! insert_selector {
+            ($selector: expr) => {
+                self.selectors.push($selector);
+                self.mtf_freq[$selector as usize] += 1;
+            };
+        }
+
+        macro_rules! insert_pending_zeros {
+            () => {
+                if index_zero_freq > 0 {
+                    index_zero_freq -= 1;
+
+                    loop {
+                        if (index_zero_freq & 1) == 1 {
+                            insert_selector!(Self::RUNB);
+                        } else {
+                            insert_selector!(Self::RUNA);
+                        }
+                        if index_zero_freq < 2 {
+                            break;
+                        }
+                        index_zero_freq -= 2;
+                        index_zero_freq /= 2;
+                    }
+
+                    index_zero_freq = 0;
+                }
+            };
+        }
+
+        for i in 0..nblock {
+            let j = if ptr[i] < 1 {
+                ptr[i] + nblock - 1
+            } else {
+                ptr[i] - 1
+            };
+
+            let byte = buf[j];
+            let index = index_selector[byte as usize];
+
+            if live_mtf[0] == index {
+                index_zero_freq += 1;
+                continue;
+            }
+
+            insert_pending_zeros!();
+
+            println!("{:?}", live_mtf);
+
+            // TODO: nightly has rotate_right?
+            // [0..loc] transformed to [loc; 0..loc-1]
+            let mut loc = 1;
+            let mut assignee = live_mtf[loc - 1];
+            while live_mtf[loc] != index {
+                let prev_assignee = assignee;
+                assignee = live_mtf[loc];
+                live_mtf[loc] = prev_assignee;
+                loc += 1;
+            }
```

Comment:
I'm not sure this would work with `rotate_right`, this is how the decoder is currently implemented:

https://github.com/paolobarbolini/bzip2-rs/blob/77fc1587d5c5a36031704a6340cfc9f47abb8c1a/src/move_to_front.rs#L22-L29

Reply:
Is this faster than just manually looping like the way I did? I think so because ig it would skip a lot of bounds checking. Also, it's more succinct.

Reply:
Actually forget it, I must have done it wrong when I tried doing it with `rotate_right` in the decoder. It seems doable with `rotate_right`.

Reply:
:thinking: rotate_right doesn't seem to optimize well https://rust.godbolt.org/z/7snz4K